code/game/machinery/autolathe/matterforge.dm

```diff
@@ -0,0 +1,606 @@
+	var/show_category
+	var/list/categories
+
+	var/list/lst = list()
+
+	var/working = FALSE
+	var/paused = FALSE
+	var/error
+	var/progress = 0
+
+	var/datum/design/current_design
+	var/list/queue = list()
+	var/queue_max = 8
+
+	var/list/disk_list = list(
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/adv_tools,
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/components,
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/circuits,
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/conveyors,
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/devices,
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/misc,
+	/obj/item/weapon/computer_hardware/hard_drive/portable/design/tools,
+	)
+	var/list/design_list = list()
+	var/speed = 2
+	var/mat_efficiency = 1
+
+	var/have_disk = FALSE
+	var/have_reagents = FALSE
+	var/have_materials = TRUE
+	var/have_design_selector = TRUE
+
+	var/list/unsuitable_materials = list(MATERIAL_BIOMATTER)
+
+	var/list/special_actions
+
+	var/global/list/error_messages = list(
+		ERR_NOTFOUND = "Design data not found.",
+		ERR_NOMATERIAL = "Not enough materials.",
+		ERR_NOREAGENT = "Not enough reagents.",
+		ERR_PAUSED = "**Construction Paused**"
+	)
+	var/tmp/obj/effect/flicker_overlay/image_load
+	var/tmp/obj/effect/flicker_overlay/image_load_material
+
+/obj/machinery/matter_nanoforge/proc/get_designs()
+	for (var/f in disk_list)
+		design_list.Add(find_files_by_type(f))
+
+/obj/machinery/matter_nanoforge/proc/find_files_by_type(typepath)
+	var/list/files = list()	
+	var/obj/item/weapon/computer_hardware/hard_drive/portable/design/c = new typepath
+	for(var/f in c.designs)
+		var/datum/design/design_object = SSresearch.get_design(f)
+		var/total_mat = 0 
+		var/list/ui_mats = design_object.ui_data["materials"]
+		var/saved_mat = ui_mats[1]
+		for(var/req_mats in ui_mats)
+			total_mat += req_mats["req"]
+		for(var/dmat in design_object.materials)
+			total_mat = total_mat +  ((1 - lst[dmat]) * 10) * 2
+		saved_mat["req"] = total_mat
+		saved_mat["name"] = MATERIAL_COMPRESSED_MATTER
+		ui_mats = list(1)
+		ui_mats[1] = saved_mat
+		design_object.ui_data["materials"] = ui_mats
+		files.Add(design_object)
+	qdel(c)
+	return files
+
+/obj/machinery/matter_nanoforge/proc/materials_data()
+	var/list/data = list()
+	data["mat_efficiency"] = mat_efficiency
+	data["mat_capacity"] = storage_capacity
+
+	var/list/M = list()
+	for(var/mtype in stored_material)
+		if(stored_material[mtype] <= 0)
+			continue
+
+		var/material/MAT = get_material_by_name(mtype)
+		var/list/ME = list("name" = MAT.display_name, "id" = mtype, "amount" = stored_material[mtype], "ejectable" = !!MAT.stack_type)
+
+		M.Add(list(ME))
+
+	data["materials"] = M
+
+	return data
+
+/obj/machinery/matter_nanoforge/ui_data()
+	var/list/data = list()
+
+	data["have_materials"] = have_materials
+	data["have_design_selector"] = have_design_selector
+
+	data["error"] = error
```

Comment:
why?

Reply:
if we have any errors

Reply:
so why is there no reference to this variable in the .tmpl file?